Incorrect string-merge results -mcmodel=large is used with clang

When `program_name_29___cpp_integration_cc__` is compiled with `clang` rather than `gcc`, it fails. Apparently all the merged strings end up as the string `ELF`. See https://github.com/davidlattimore/wild/pull/140. We should probably first add an option to allow overriding the compiler - i.e. #179.

If I edit the source file test arguments to be:
```
//#Config:model-large:default
//#CompArgs:-mcmodel=large
//#LinkArgs:--cc=clang -Wl,-z,now
```

putting clang instead of gcc, the test fails thus:

```
thread 'main' panicked at /home/andrew/workspace/wild/wild/tests/sources/rust-integration.rs:21:5:
Make sure unwinding works
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
thread 'main' panicked at /home/andrew/workspace/wild/wild/tests/sources/rust-integration.rs:21:5:
Make sure unwinding works
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
test integration_test::program_name_28___rust_integration_dynamic_rs__ ... ok
/usr/bin/ld: /home/andrew/workspace/wild/wild/tests/build/cpp-integration.model-large-8ead9704af940f25.o: in function `main':
cpp-integration.cc:(.text+0x40): undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::basic_string()'
/usr/bin/ld: cpp-integration.cc:(.text+0x6a): undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::operator+=(char const*)'
/usr/bin/ld: cpp-integration.cc:(.text+0x9b): undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::operator+=(char const*)'
/usr/bin/ld: cpp-integration.cc:(.text+0xc5): undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::operator+=(char const*)'
/usr/bin/ld: cpp-integration.cc:(.text+0xef): undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::operator+=(char const*)'
/usr/bin/ld: cpp-integration.cc:(.text+0x120): undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::operator+=(char const*)'
/usr/bin/ld: /home/andrew/workspace/wild/wild/tests/build/cpp-integration.model-large-8ead9704af940f25.o:cpp-integration.cc:(.text+0x14a): more undefined references to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::operator+=(char const*)' follow
/usr/bin/ld: /home/andrew/workspace/wild/wild/tests/build/cpp-integration.model-large-8ead9704af940f25.o: in function `main':
cpp-integration.cc:(.text+0x1b6): undefined reference to `std::cout'
/usr/bin/ld: cpp-integration.cc:(.text+0x1ca): undefined reference to `std::basic_ostream<char, std::char_traits<char> >& std::operator<< <char, std::char_traits<char>, std::allocator<char> >(std::basic_ostream<char, std::char_traits<char> >&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)'
/usr/bin/ld: cpp-integration.cc:(.text+0x1d9): undefined reference to `std::basic_ostream<char, std::char_traits<char> >& std::endl<char, std::char_traits<char> >(std::basic_ostream<char, std::char_traits<char> >&)'
/usr/bin/ld: cpp-integration.cc:(.text+0x1f0): undefined reference to `std::ostream::operator<<(std::ostream& (*)(std::ostream&))'
/usr/bin/ld: cpp-integration.cc:(.text+0x217): undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::~basic_string()'
/usr/bin/ld: cpp-integration.cc:(.text+0x24b): undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::~basic_string()'
/usr/bin/ld: /home/andrew/workspace/wild/wild/tests/build/cpp-integration.model-large-8ead9704af940f25.o: in function `bool std::operator==<char, std::char_traits<char>, std::allocator<char> >(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, char const*)':
cpp-integration.cc:(.text._ZSteqIcSt11char_traitsIcESaIcEEbRKNSt7__cxx1112basic_stringIT_T0_T1_EEPKS5_[_ZSteqIcSt11char_traitsIcESaIcEEbRKNSt7__cxx1112basic_stringIT_T0_T1_EEPKS5_]+0x39): undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::size() const'
/usr/bin/ld: cpp-integration.cc:(.text._ZSteqIcSt11char_traitsIcESaIcEEbRKNSt7__cxx1112basic_stringIT_T0_T1_EEPKS5_[_ZSteqIcSt11char_traitsIcESaIcEEbRKNSt7__cxx1112basic_stringIT_T0_T1_EEPKS5_]+0x71): undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::size() const'
/usr/bin/ld: cpp-integration.cc:(.text._ZSteqIcSt11char_traitsIcESaIcEEbRKNSt7__cxx1112basic_stringIT_T0_T1_EEPKS5_[_ZSteqIcSt11char_traitsIcESaIcEEbRKNSt7__cxx1112basic_stringIT_T0_T1_EEPKS5_]+0x8d): undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::data() const'
/usr/bin/ld: /home/andrew/workspace/wild/wild/tests/build/cpp-integration.model-large-8ead9704af940f25.o:(.data.rel.local.DW.ref.__gxx_personality_v0[DW.ref.__gxx_personality_v0]+0x0): undefined reference to `__gxx_personality_v0'
clang: error: linker command failed with exit code 1 (use -v to see invocation)
test integration_test::program_name_29___cpp_integration_cc__ ... FAILED
thread 'main' panicked at /home/andrew/workspace/wild/wild/tests/sources/rust-integration.rs:21:5:
Make sure unwinding works
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
thread 'main' panicked at /home/andrew/workspace/wild/wild/tests/sources/rust-integration.rs:21:5:
Make sure unwinding works
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
test integration_test::program_name_27___rust_integration_rs__ ... ok
test integration_test::program_name_26___libc_integration_c__ ... ok
test integration_test::program_name_30___rust_tls_rs__ ... ok

failures:

---- integration_test::program_name_29___cpp_integration_cc__ stdout ----
Error: Failed to build program `cpp-integration.cc` with linker `ld` config `model-large`

Caused by:
    Linker failed. Relink with:
    clang --ld-path=/usr/bin/ld -Wl,-z,now -o /home/andrew/workspace/wild/wild/tests/build/cpp-integration.cc-model-large.ld /home/andrew/workspace/wild/wild/tests/build/cpp-integration.model-large-8ead9704af940f25.o /home/andrew/workspace/wild/wild/tests/build/cpp-integration-2.model-large-d2a94c574f4beb01.o


failures:
    integration_test::program_name_29___cpp_integration_cc__

test result: FAILED. 29 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 3.53s

error: test failed, to rerun pass `-p wild --test integration_tests`

```


Would they be a reproducer for this issue, to allow me to tinker with it and see what the cause it?

--- a/libwild/src/elf_writer.rs
+++ b/libwild/src/elf_writer.rs
@@ -1917,7 +1917,13 @@ fn apply_relocation<S: StorageModel, A: Arch>(
             .bitand(mask.got_entry)
             .wrapping_add(addend as u64),
         RelocationKind::SymRelGotBase => resolution
-            .value()
+            .value_with_addend(
+                addend,
+                symbol_index,
+                object_layout,
+                &layout.merged_strings,
+                &layout.merged_string_start_addresses,
+            )?
             .bitand(mask.symbol_plus_addend)
             .wrapping_sub(layout.got_base().bitand(mask.got)),
         RelocationKind::PltRelGotBase => resolution

--- a/linker-diff/src/lib.rs
+++ b/linker-diff/src/lib.rs
@@ -158,6 +158,8 @@ impl Config {
                 "section.bss.alignment",
                 "section.gnu.build.attributes",
                 "section.annobin.notes.entsize",
+                // We don't yet group .lrodata sections separately.
+                "section.lrodata",
                 // We currently output version info when linking against the interpreter
                 // (ld-linux-x86-64.so.2). GNU ld doesn't.
                 ".dynamic.DT_VERNEEDNUM",

--- a/wild/tests/sources/cpp-integration.cc
+++ b/wild/tests/sources/cpp-integration.cc
@@ -25,8 +25,14 @@
 //#CompArgs:-mcmodel=large
 //#LinkArgs:--cc=g++ -Wl,-z,now
 //#EnableLinker:lld
-//#Cross:false
 // TODO: Ubuntu: cc1plus: sorry, unimplemented: code model 'large' with '-fPIC'
+//#Arch: x86_64
+
+//#Config:clang-model-large:default
+//#Compiler:clang
+//#CompArgs:-mcmodel=large
+//#LinkArgs:--cc=clang++ -Wl,-z,now
+//#EnableLinker:lld
 //#Arch: x86_64
 
 #include <iostream>

